File: ios/Classes/FlutoryxUploaderPlugin.swift
import Flutter
import UIKit

public class FlutoryxUploaderPlugin: NSObject, FlutterPlugin, FlutterStreamHandler, FlutterApplicationLifeCycleDelegate {
  
  private var eventSink: FlutterEventSink?
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutoryx_uploader", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "flutoryx_uploader/events", binaryMessenger: registrar.messenger())
    
    let instance = FlutoryxUploaderPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    eventChannel.setStreamHandler(instance)
    registrar.addApplicationDelegate(instance)
    
    // Initialize UploadManager to reconnect session and resume tasks
    _ = UploadManager.shared
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "uploadFile":
        handleUploadFile(call, result: result)
    case "uploadFiles":
        handleUploadFiles(call, result: result)
    case "cancelUpload":
        if let args = call.arguments as? [String: Any], let taskId = args["taskId"] as? String {
            PersistenceManager.shared.updateTaskStatus(taskId: taskId, status: .canceled)
            result(nil)
        } else {
             result(FlutterError(code: "INVALID_ARGUMENT", message: "taskId missing", details: nil))
        }
    case "pauseUpload":
         if let args = call.arguments as? [String: Any], let taskId = args["taskId"] as? String {
            PersistenceManager.shared.updateTaskStatus(taskId: taskId, status: .paused)
             // UploadManager logic to pause?
             // Since chunks are atomic, we just stop scheduling next chunk.
            result(nil)
        } else {
             result(FlutterError(code: "INVALID_ARGUMENT", message: "taskId missing", details: nil))
        }
    case "resumeUpload":
          if let args = call.arguments as? [String: Any], let taskId = args["taskId"] as? String {
            UploadManager.shared.startUpload(taskId: taskId)
            result(nil)
             result(FlutterError(code: "INVALID_ARGUMENT", message: "taskId missing", details: nil))
        }
    case "getTasks":
        let tasks = PersistenceManager.shared.getAllTasks()
        let mappedTasks = tasks.map { task -> [String: Any] in
            return [
                "taskId": task.taskId,
                "status": task.status.rawValue.uppercased(),
                "progress": task.progress,
                "speed": task.speed,
                "eta": task.eta ?? NSNull(),
                "errorMessage": NSNull()
            ]
        }
        result(mappedTasks)
    case "removeTask":
        if let args = call.arguments as? [String: Any], let taskId = args["taskId"] as? String {
            UploadManager.shared.cancelUpload(taskId: taskId)
            PersistenceManager.shared.removeTask(taskId: taskId)
            result(nil)
        } else {
             result(FlutterError(code: "INVALID_ARGUMENT", message: "taskId missing", details: nil))
        }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  private func handleUploadFile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      guard let args = call.arguments as? [String: Any],
            let filePath = args["filePath"] as? String,
            let endpoint = args["endpoint"] as? String,
            let config = args["config"] as? [String: Any] else {
          result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing arguments", details: nil))
          return
      }
      
      let taskId = UUID().uuidString
      let chunkSize = config["chunkSize"] as? Int ?? 1024 * 1024
      let maxParallel = config["maxParallelUploads"] as? Int ?? 2
      let adaptive = config["adaptiveNetwork"] as? Bool ?? true
      let showNotification = config["showNotification"] as? Bool ?? true
      let uploadModeStr = config["uploadMode"] as? String ?? "direct"
      let uploadMode = UploadMode(rawValue: uploadModeStr) ?? .direct
      
      let headers = args["headers"] as? [String: String]
      let data = args["data"] as? [String: String]
      
      let task = UploadTask(
        taskId: taskId,
        filePath: filePath,
        endpoint: endpoint,
        chunkSize: chunkSize,
        maxParallelUploads: maxParallel,
        adaptiveNetwork: adaptive,
        showNotification: showNotification,
        headers: headers,
        data: data,
        status: .enqueued,
        progress: 0,
        uploadedChunkIndices: [],
        startTime: Date(),
        speed: 0,
        eta: nil,
        uploadMode: uploadMode
      )
      
      PersistenceManager.shared.addTask(task)
      UploadManager.shared.startUpload(taskId: taskId)
      
      result(taskId)
  }
  
  private func handleUploadFiles(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
       guard let args = call.arguments as? [String: Any],
            let filePaths = args["filePaths"] as? [String],
            let endpoint = args["endpoint"] as? String,
            let config = args["config"] as? [String: Any] else {
          result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing arguments", details: nil))
          return
      }
      
      var taskIds: [String] = []
      
       let chunkSize = config["chunkSize"] as? Int ?? 1024 * 1024
      let maxParallel = config["maxParallelUploads"] as? Int ?? 2
      let adaptive = config["adaptiveNetwork"] as? Bool ?? true
      let showNotification = config["showNotification"] as? Bool ?? true
      let uploadModeStr = config["uploadMode"] as? String ?? "direct"
      let uploadMode = UploadMode(rawValue: uploadModeStr) ?? .direct
      
      let headers = args["headers"] as? [String: String]
      let data = args["data"] as? [String: String]
      
      for path in filePaths {
          let taskId = UUID().uuidString
          taskIds.append(taskId)
          
           let task = UploadTask(
                taskId: taskId,
                filePath: path,
                endpoint: endpoint,
                chunkSize: chunkSize,
                maxParallelUploads: maxParallel,
                adaptiveNetwork: adaptive,
                showNotification: showNotification,
                headers: headers,
                data: data,
                status: .enqueued,
                progress: 0,
                uploadedChunkIndices: [],
                startTime: Date(),
                speed: 0,
                eta: nil,
                uploadMode: uploadMode
              )
           PersistenceManager.shared.addTask(task)
           UploadManager.shared.startUpload(taskId: taskId)
      }
      
      result(taskIds)
  }
    
  // MARK: - Event Channel
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
      self.eventSink = events
      NotificationCenter.default.addObserver(self, selector: #selector(onProgress(_:)), name: NSNotification.Name("FlutoryxProgress"), object: nil)
      return nil
  }
    
  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
      self.eventSink = nil
      NotificationCenter.default.removeObserver(self)
      return nil
  }
    
  @objc private func onProgress(_ notification: Notification) {
      guard let userInfo = notification.userInfo, let eventSink = self.eventSink else { return }
      eventSink(userInfo)
  }
    
  // MARK: - Application Delegate
  public func application(_ application: UIApplication, handleEventsForBackgroundURLSession identifier: String, completionHandler: @escaping () -> Void) -> Bool {
      // Pass completion handler to UploadManager
      // Note: check identifier == "com.flutoryx.uploader.background"
      if identifier == "com.flutoryx.uploader.background" {
          UploadManager.shared.backgroundCompletionHandler = completionHandler
          return true
      }
      return false
  }
}

File: android/src/main/kotlin/com/example/flutoryx_uploader/FlutoryxUploaderPlugin.kt
package com.example.flutoryx_uploader

import android.content.Context
import androidx.work.BackoffPolicy
import androidx.work.Constraints
import androidx.work.Data
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import androidx.work.WorkInfo
import com.example.flutoryx_uploader.db.UploadDatabase
import com.example.flutoryx_uploader.db.UploadTaskEntity
import com.google.gson.Gson
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.UUID
import java.util.concurrent.TimeUnit

/** FlutoryxUploaderPlugin */
class FlutoryxUploaderPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  private lateinit var channel : MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context
  private lateinit var db: UploadDatabase
  private lateinit var workManager: WorkManager
  private val mainScope = CoroutineScope(Dispatchers.Main)
  private var eventSink: EventChannel.EventSink? = null
  private var progressJob: Job? = null

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutoryx_uploader")
    channel.setMethodCallHandler(this)
    
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutoryx_uploader/events")
    eventChannel.setStreamHandler(this)

    db = UploadDatabase.getDatabase(context)
    workManager = WorkManager.getInstance(context)
  
    startProgressWatcher()
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "uploadFile" -> {
        val taskId = UUID.randomUUID().toString()
        handleUploadFile(call, result, taskId)
        result.success(taskId)
      }
      "uploadFiles" -> {
        handleUploadFiles(call, result)
      }
      "cancelUpload" -> {
        val taskId = call.argument<String>("taskId")
        if (taskId != null) {
          mainScope.launch {
            cancelUpload(taskId)
            result.success(null)
          }
        } else {
          result.error("INVALID_ARGUMENT", "taskId is null", null)
        }
      }
      "pauseUpload" -> {
          val taskId = call.argument<String>("taskId")
           if (taskId != null) {
               mainScope.launch {
                   pauseUpload(taskId)
                   result.success(null)
               }
           } else {
               result.error("INVALID_ARGUMENT", "taskId is null", null)
           }
      }
      "resumeUpload" -> {
          val taskId = call.argument<String>("taskId")
          if (taskId != null) {
              mainScope.launch {
                  resumeUpload(taskId)
                   result.success(null)
              }
          } else {
              result.error("INVALID_ARGUMENT", "taskId is null", null)
          }
      }
      "getTasks" -> {
          mainScope.launch {
              val tasks = withContext(Dispatchers.IO) {
                  db.uploadDao().getAllTasks()
              }
              val mappedTasks = tasks.map { task ->
                  mapOf(
                      "taskId" to task.taskId,
                      "status" to task.status,
                      "progress" to task.progress,
                      "speed" to task.speed,
                      "eta" to task.eta,
                      "errorMessage" to null // TODO: Add error column to DB
                  )
              }
              result.success(mappedTasks)
          }
      }
      "removeTask" -> {
          val taskId = call.argument<String>("taskId")
          if (taskId != null) {
              mainScope.launch {
                  // Cancel WorkManager task by Unique Name (which is the taskId)
                  workManager.cancelUniqueWork(taskId)
                  
                  // Delete from DB
                   withContext(Dispatchers.IO) {
                       db.uploadDao().delete(taskId)
                   }
                  result.success(null)
              }
          } else {
              result.error("INVALID_ARGUMENT", "taskId is null", null)
          }
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun handleUploadFile(call: MethodCall, result: Result?, taskId: String) {
    val filePath = call.argument<String>("filePath")!!
    val endpoint = call.argument<String>("endpoint")!!
    val configMap = call.argument<Map<String, Any>>("config")!!
    val headers = call.argument<Map<String, String>>("headers")
    val data = call.argument<Map<String, String>>("data")

    val gson = Gson()
    val headersJson = if (headers != null) gson.toJson(headers) else null
    val dataJson = if (data != null) gson.toJson(data) else null

    val chunkSize = (configMap["chunkSize"] as? Number)?.toInt() ?: 1024 * 1024
    val maxParallelUploads = (configMap["maxParallelUploads"] as? Number)?.toInt() ?: 2
    val adaptiveNetwork = configMap["adaptiveNetwork"] as? Boolean ?: true
    val showNotification = configMap["showNotification"] as? Boolean ?: true
    val uploadMode = configMap["uploadMode"] as? String ?: "direct" // Default to direct
    
    // retry logic extraction...

    val taskEntity = UploadTaskEntity(
      taskId = taskId,
      filePath = filePath,
      endpoint = endpoint,
      chunkSize = chunkSize,
      maxParallelUploads = maxParallelUploads,
      adaptiveNetwork = adaptiveNetwork,
      maxRetries = 3, // TODO: extract from config
      status = UploadStatus.ENQUEUED.toResultString(),
      showNotification = showNotification,
      startTime = System.currentTimeMillis(),
      headersJson = headersJson,
      dataJson = dataJson,
      uploadMode = uploadMode
    )

    mainScope.launch {
       withContext(Dispatchers.IO) {
           db.uploadDao().insert(taskEntity)
       }
       enqueueWork(taskId, adaptiveNetwork, uploadMode)
    }
  }
  
  private fun handleUploadFiles(call: MethodCall, result: Result) {
      val filePaths = call.argument<List<String>>("filePaths")!!
      val endpoint = call.argument<String>("endpoint")!!
      // reuse same config for all
      
      val taskIds = mutableListOf<String>()
      
      filePaths.forEach { path ->
          val taskId = UUID.randomUUID().toString()
          taskIds.add(taskId)
          
          // Re-create arguments for this specific file
          val singleFileArgs = HashMap(call.arguments as Map<String, Any>)
          singleFileArgs["filePath"] = path
          
          val methodCall = MethodCall("uploadFile", singleFileArgs)
          handleUploadFile(methodCall, null, taskId)
      }
      result.success(taskIds)
  }

  private fun enqueueWork(taskId: String, adaptiveNetwork: Boolean, uploadMode: String) {
      val constraints = Constraints.Builder()
          .setRequiredNetworkType(NetworkType.CONNECTED)
          .build()

      val data = Data.Builder()
          .putString("taskId", taskId)
          .build()

      val workerClass = if (uploadMode == "chunked") UploadWorker::class.java else DirectUploadWorker::class.java

      val request = OneTimeWorkRequest.Builder(workerClass)
          .setConstraints(constraints)
          .setInputData(data)
          .addTag(taskId)
          .setBackoffCriteria(BackoffPolicy.LINEAR, 10, TimeUnit.SECONDS)
          .build()

      workManager.enqueueUniqueWork(taskId, ExistingWorkPolicy.REPLACE, request)
  }
  
  private suspend fun cancelUpload(taskId: String) {
      workManager.cancelUniqueWork(taskId)
      withContext(Dispatchers.IO) {
          db.uploadDao().updateStatus(taskId, UploadStatus.CANCELED.toResultString())
      }
  }

    private suspend fun pauseUpload(taskId: String) {
        workManager.cancelUniqueWork(taskId) // Cancels current worker
        withContext(Dispatchers.IO) {
            db.uploadDao().updateStatus(taskId, UploadStatus.PAUSED.toResultString())
        }
    }
    
    private suspend fun resumeUpload(taskId: String) {
          withContext(Dispatchers.IO) {
            val task = db.uploadDao().getTask(taskId)
            if (task != null) {
                db.uploadDao().updateStatus(taskId, UploadStatus.ENQUEUED.toResultString())
                enqueueWork(taskId, task.adaptiveNetwork, task.uploadMode)
            }
        }
    }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
    progressJob?.cancel()
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }
  
  // Polling DB for progress updates. 
  // Ideally we use Room's Flow or LiveData, but we need to pipe it to EventChannel.
  // A simple poller is easiest for now to avoid extensive boilerplate with Flows -> EventChannel.
  private fun startProgressWatcher() {
      progressJob = mainScope.launch {
          while (isActive) {
              val tasks = withContext(Dispatchers.IO) {
                  db.uploadDao().getAllPendingTasks() // Maybe get all active tasks
              }
              // Actually we want ALL tasks that might have updates.
              // For efficiency, maybe only grab RUNNING ones or ones updated recently.
              // But simpler: just grab all RUNNING/ENQUEUED/FAILED/COMPLETED that changed?
              // Let's just grab active tasks.
              
              // To properly stream 'progress', we need to check ALL tasks that are relevant. 
              // But iterating whole DB is bad.
              // Let's rely on the Worker updating the DB, and we only poll tasks that are NOT in a final state?
              // Or user specifically subscribes to ID? The API is a global stream.
              
              // Let's polling active tasks.
               val activeTasks = withContext(Dispatchers.IO) {
                  // Query is predefined in DAO: "SELECT * FROM upload_tasks WHERE status = 'ENQUEUED' OR status = 'RUNNING'"
                  // Add PAUSED?
                  db.uploadDao().getAllTasks()
                }
               
               for (task in activeTasks) {
                   val event = mapOf(
                       "taskId" to task.taskId,
                       "status" to task.status,
                       "progress" to task.progress,
                       "speed" to task.speed,
                       "eta" to task.eta,
                       "errorMessage" to null // TODO store error message
                   )
                   eventSink?.success(event)
               }
               
              delay(500) // 500ms poll
          }
      }
  }
}
